src/transformers/models/vibevoice_semantic_tokenizer/modular_vibevoice_semantic_tokenizer.py

```diff
@@ -0,0 +1,390 @@
+        self.activation = ACT2FN[config.hidden_act]
+        self.linear2 = nn.Linear(config.ffn_expansion * hidden_size, hidden_size, bias=config.bias)
+
+    def forward(self, hidden_states):
+        return self.linear2(self.activation(self.linear1(hidden_states)))
+
+
+class VibeVoiceConv1dCache:
+    """
+    Similar to Mimi's Cache: https://github.com/huggingface/transformers/blob/cad7eeeb5e8a173f8d7d746ccdb6ef670ffe6be4/src/transformers/models/mimi/modeling_mimi.py#L76
+    But with `batch_mask` support for selective cache updates.
+
+    Original (uses unique key per layer and sample): https://github.com/pengzhiliang/transformers/blob/6e6e60fb95ca908feb0b039483adcc009809f579/src/transformers/models/vibevoice/modular_vibevoice_tokenizer.py#L174
+    """
+
+    def __init__(self, num_layers: int):
+        self.cache = [None] * num_layers
+
+    def update(
+        self,
+        layer_idx: int,
+        context: int,
+        hidden_states: torch.Tensor,
+        batch_mask: Optional[torch.Tensor] = None,
+    ):
+        """
+        Updates the padding cache with the new padding states for the layer `layer_idx` and returns the current cache.
+        Similar to Mimi's update method.
+
+        Parameters:
+            layer_idx (`int`):
+                The index of the layer to cache the states for.
+            hidden_states (`torch.Tensor`):
+                The hidden states to be partially cached.
+            batch_mask (`torch.LongTensor` of shape `(batch_size,)`, *optional*):
+                Indices of samples to update cache for.
+            context (`int`):
+                The amount of context for this layer.
+
+        Returns:
+            `torch.Tensor`, the current padding cache for the specified samples.
+        """
+        batch_size, channels, _ = hidden_states.shape
+
+        if batch_mask is None:
+            batch_mask = torch.arange(batch_size, device=hidden_states.device)
+        else:
+            if len(batch_mask) != batch_size:
+                raise ValueError("batch_mask length must match batch size")
+
+        # Get existing cache or initialize if None
+        existing_cache = self.cache[layer_idx]
+        if existing_cache is None:
+            current_cache = torch.zeros(
+                batch_size, channels, max(context, 0), device=hidden_states.device, dtype=hidden_states.dtype
+            )
+        else:
+            current_cache = existing_cache[batch_mask]
+
+        # Update the cache with full input as in original (otherwise not enough context):
+        # https://github.com/pengzhiliang/transformers/blob/6e6e60fb95ca908feb0b039483adcc009809f579/src/transformers/models/vibevoice/modular_vibevoice_tokenizer.py#L345 
+        input_with_context = torch.cat([current_cache, hidden_states], dim=-1)
+        if context > 0:
+            new_padding_states = input_with_context[..., -context:]
```

Comment:
Left comments like this to point out difference with Mimi,.

In this case, Mimi would only take context from hidden states ([here](https://github.com/huggingface/transformers/blob/cad7eeeb5e8a173f8d7d746ccdb6ef670ffe6be4/src/transformers/models/mimi/modeling_mimi.py#L153)) but this breaks for VibeVoice (hidden_states is not large enough)